### ISSUE TITLE
Can't fork fibers and propagate them outside zipWithPar

### ISSUE THREAD

**Issue by iravid** [zio/zio#3073]
```scala
val printer =
  UIO(println("ok")).repeat(Schedule.fixed(1.second))

(printer.fork <& ZIO.unit).flatMap(_.join)
```

This fails immediately because `printer` is interrupted after `zipWithPar` ends. It is interrupted because `zipWithPar` races both sides in a new fiber, which interrupts all descendants when it ends.

cc @jdegoes @adamgfraser

**Comment by iravid** [comment]
refs https://github.com/zio/zio-kafka/issues/135

**Comment by iravid** [comment]
@jdegoes Some possible ideas I had:
- add a regional combinator that forks fiber under a different parent:
```
def forkOnParent(zio: ZIO[R, E, A]): ZIO[R, E, A]

// Moves children one level up the hierarchy
forkOnParent(UIO(println("ok").repeat(Schedule.fixed(1.second)).fork).fork.flatMap(_.join)
```
- add a forking behavior that says that when the parent ends, the fibers await the termination of their grandparent

**Comment by jdegoes** [comment]
How about:

```scala
ZIO.supervisor(supervisor => ... )
...
trait Supervisor {
  def fork[R, E, A](zio: ZIO[R, E, A]): ZIO[R, Nothing, Fiber[E, A]]
}
```

Then we can grab the supervisor at the beginning and pass it down all the way to the fibers forked with `raceWith`. Would that work?

Tricky, almost requires `evalOn` to do properly.

Essentially we have to somewhat disentangle the notion of "scope" from "fiber".

**Comment by iravid** [comment]
I'm not sure it's enough to have something that directly forks an effect. You need something that modifies where fibers are forked throughout an entire region to make this composable.

So it'd have to be:
```scala
ZIO.supervisorRestore(restore => )
trait Supervisor {
  def apply[R, E, A](zio: ZIO[R, E, A]): ZIO[R, E, A]
}
```

**Comment by adamgfraser** [comment]
I think having the parent adopt is interesting but we would need to remove the fiber from the parent and add it to the grandparent in way that was concurrency safe. I think we would also need to check if the grandparent was itself in wind down and if so transplant to it’s parent or possibly the root fibers.

**Comment by jdegoes** [comment]
@iravid Let me think about that. It's close...

@adamgfraser I went down that path before I deleted transplanting / relocation. It's super complex because we can't atomically operate across the state of more than one fiber.

**Comment by adamgfraser** [comment]
@jdegoes Yes it seems like you would need something like STM to do that.

**Comment by adamgfraser** [comment]
If we just fork on the grandparent are we creating another set of issues where the child is not going to be interrupted if the child is interrupted?

**Comment by adamgfraser** [comment]
Resolves by #3829.
